### PR TITLE
Start implementing the event auth checks

### PR DESCRIPTION
**Pull request by NegativeMjark**

### CODE REVIEW

File: matrixeventauth/eventauth.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
+package matrixeventauth
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+)
+
+// A Event has the fields necessary to authenticate a matrix event.
```
Comment (Kegsay) [inline matrixeventauth/eventauth.go]: An

File: matrixeventauth/eventauth.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+// It can be unmarshalled from the event JSON.
+type Event struct {
+	RoomID     string              `json:"room_id"`
+	EventID    string              `json:"event_id"`
+	Sender     string              `json:"sender"`
+	Type       string              `json:"type"`
+	StateKey   *string             `json:"state_key"`
+	Content    json.RawMessage     `json:"content"`
+	PrevEvents [][]json.RawMessage `json:"prev_events"`
+	Redacts    string              `json:"redacts"`
```
Comment (Kegsay) [inline matrixeventauth/eventauth.go]: No `unsigned`?

Comment (NegativeMjark) [inline matrixeventauth/eventauth.go]: You don't need the unsigned key to authenticate an event.

File: matrixeventauth/eventauth.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+		case "m.room.aliases":
+			// Alias events need:
+			//  * The create event.
+			//    https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L128
+			// Alias events need no further authentication.
+			// https://github.com/matrix-org/synapse/blob/v0.18.5/synapse/api/auth.py#L160
+			result.Create = true
+		case "m.room.member":
+			// Member events need:
+			//  * The previous membership of the target.
```
Comment (Kegsay) [inline matrixeventauth/eventauth.go]: Where do we get this information in the code?

Comment (NegativeMjark) [inline matrixeventauth/eventauth.go]: from the:

```members = append(members, event.Sender, *event.StateKey)```

Comment (Kegsay) [inline matrixeventauth/eventauth.go]: But this isn't the *previous* membership?

File: eventauth/eventauth.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+	// Is the m.room.power_levels event needed to auth the event.
+	PowerLevels bool
+	// List of m.room.member state_keys needed to auth the event
+	Member []string
+	// List of m.room.third_party_invite state_keys
+	ThirdPartyInvite []string
+}
+
+// StateNeededForAuth returns the event types and state_keys needed to authenticate an event.
+func StateNeededForAuth(events []Event) (result StateNeeded) {
```
Comment (Kegsay) [inline eventauth/eventauth.go]: What are `events`? Why isn't this just 1 event?

File: eventauth/eventauth.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+// memberContent is the JSON content of a m.room.member event.
+type memberContent struct {
+	Membership       string          `json:"membership"`
+	ThirdPartyInvite json.RawMessage `json:"third_party_invite"`
+}
+
+// Remove duplicate items from a sorted list.
+// Takes the same interface as sort.Sort
+// Returns the length of the date without duplicates
```
Comment (Kegsay) [inline eventauth/eventauth.go]: `date`

**Comment by Kegsay** [comment]
Otherwise LGTM

**Comment by NegativeMjark** [comment]
@Kegsay I've added some tests PTAL.

**Comment by Kegsay** [comment]
Beautiful. LGTM

**Pull request merged**